src/TestStream.cpp

```diff
@@ -17,35 +17,89 @@ void TestStream::pushDataToDriver(vector<uint8_t> const& data)
 vector<uint8_t> TestStream::readDataFromDriver()
 {
     vector<uint8_t> temp;
-    temp.swap(from_device);
+    temp.swap(from_driver);
     return temp;
 }
 
 void TestStream::waitRead(base::Time const& timeout)
 {
-    if (to_device.empty())
+    if (to_driver.empty())
         throw TimeoutError(TimeoutError::NONE, "no data in to_device");
 }
 void TestStream::waitWrite(base::Time const& timeout)
 {
 }
 
+void TestStream::EXPECT_REPLY(std::vector<uint8_t> const& expectation, std::vector<uint8_t> const& reply)
+{
+    if(!mock_mode)
+        throw MockContextException();
+    std::vector<uint8_t> exp;
+    exp.insert(exp.end(),expectation.begin(),expectation.end());
+    std::vector<uint8_t> rep = reply;
+    expectations.push_back(exp);
+    replies.push_back(reply);
+}
+
+
 size_t TestStream::read(uint8_t* buffer, size_t buffer_size)
 {
-    size_t read_size = min(to_device.size(), buffer_size);
-    std::memcpy(buffer, to_device.data(), read_size);
-    to_device.erase(to_device.begin(), to_device.begin() + read_size);
-    return read_size;
+       size_t read_size = min(to_driver.size(), buffer_size);
+       std::memcpy(buffer, to_driver.data(), read_size);
+       to_driver.erase(to_driver.begin(), to_driver.begin() + read_size);
+       return read_size;
 }
 
 size_t TestStream::write(uint8_t const* buffer, size_t buffer_size)
 {
-    from_device.insert(from_device.end(), buffer, buffer + buffer_size);
-    return buffer_size;
+    if(mock_mode)
+    {
+        from_driver.clear();
+        from_driver.insert(from_driver.end(), buffer, buffer + buffer_size);
+        if(expectations.size() > 0)
```

Comment:
If you want to know if a container is empty or not, use `empty()`. Testing against size is both unclear when reading, and may be O(N) on e.g. lists